File: shared/src/main/java/com/openmeet/shared/data/rating/RatingDAO.java
package com.openmeet.shared.data.rating;

import com.openmeet.shared.data.storage.DAO;
import com.openmeet.shared.data.storage.GenericDAO;
import com.openmeet.shared.data.storage.SQLDAO;
import com.openmeet.shared.exceptions.InvalidPrimaryKeyException;

import javax.sql.DataSource;
import java.sql.SQLException;
import java.util.HashMap;
import java.util.List;

public class RatingDAO extends SQLDAO implements DAO<Rating> {

    public RatingDAO(DataSource source) {
        super(source);
    }

    @Override
    public List<Rating> doRetrieveByCondition(String condition) throws SQLException {
        return GenericDAO.genericDoRetrieveByCondition(Rating.RATING, condition, new RatingExtractor(), source);
    }

    @Override
    public Rating doRetrieveByKey(String key) throws SQLException, InvalidPrimaryKeyException {

        if (key == null) {
            throw new InvalidPrimaryKeyException(key);
        }

        List<Rating> rating = doRetrieveByCondition(
                String.format("%s = '%s'", Rating.RATING_ID, key)
        );

        return rating.isEmpty() ? null : rating.get(0);
    }

    @Override
    public List<Rating> doRetrieveAll() throws SQLException {
        return doRetrieveByCondition("TRUE");
    }

    @Override
    public List<Rating> doRetrieveAll(int row_count) throws SQLException {
        return doRetrieveByCondition("TRUE LIMIT " + row_count);
    }

    @Override
    public List<Rating> doRetrieveAll(int offset, int row_count) throws SQLException {
        return doRetrieveByCondition("TRUE LIMIT " + offset + ", " + row_count);
    }

    @Override
    public boolean doSave(Rating obj) throws SQLException {
        return GenericDAO.genericDoSave(Rating.RATING, obj.toHashMap(), this.source);
    }

    @Override
    public boolean doSave(HashMap<String, ?> values) throws SQLException {
        return GenericDAO.genericDoSave(Rating.RATING, values, this.source);
    }

    @Override
    public boolean doUpdate(HashMap<String, ?> values, String condition) throws SQLException {
        return GenericDAO.genericDoUpdate(Rating.RATING, condition, values, this.source);
    }

    @Override
    public boolean doSaveOrUpdate(Rating obj) throws SQLException {

        if (doRetrieveByKey(String.valueOf(obj.getId())) == null)
            return doSave(obj);

        return doUpdate(obj.toHashMap(),
                String.format("%s = '%s'", Rating.RATING_ID, obj.getId()));
    }

    @Override
    public boolean doDelete(String condition) throws SQLException {
        return GenericDAO.genericDoDelete(Rating.RATING, condition, this.source);
    }
}

File: mobileapp/app/src/main/java/com/openmeet/data/rating/RatingDAO.kt
package com.openmeet.data.rating

import com.openmeet.shared.data.storage.DAO
import com.openmeet.shared.data.storage.SQLDAO
import java.util.HashMap
import javax.sql.DataSource

class RatingDAO(source: DataSource) : SQLDAO(source), DAO<Rating> {

    override fun doRetrieveByCondition(condition: String?): MutableList<Rating> {
        return genericDoRetrieveByCondition(
            Rating.RATING,
            condition,
            RatingExtractor(),
            source
        )
    }

    override fun doRetrieveByKey(key: String?): Rating? {

        if (key == null) {
            throw IllegalArgumentException("null")
        }

        val rating = doRetrieveByCondition("$Rating.RATING_ID = '$key'")

        return if (rating.isEmpty()) null else rating[0]
    }

    override fun doRetrieveAll(): MutableList<Rating> {
        return doRetrieveByCondition("TRUE")
    }

    override fun doRetrieveAll(row_count: Int): MutableList<Rating> {
        return doRetrieveByCondition("TRUE LIMIT $row_count")
    }

    override fun doRetrieveAll(offset: Int, row_count: Int): MutableList<Rating> {
        return doRetrieveByCondition("TRUE LIMIT $offset, $row_count");
    }

    override fun doSave(obj: Rating?): Boolean {
        return genericDoSave(Rating.RATING, obj?.toHashMap(), source)
    }

    override fun doSave(values: HashMap<String, *>?): Boolean {
        return genericDoSave(Rating.RATING, values, source)
    }

    override fun doUpdate(values: HashMap<String, *>?, condition: String?): Boolean {
        return genericDoUpdate(Rating.RATING, condition, values, source)
    }

    override fun doSaveOrUpdate(obj: Rating?): Boolean {

        if (doRetrieveByKey(obj?.id.toString()) == null) {
            return doSave(obj)
        }
        return doUpdate(obj?.toHashMap(), "$Rating.RATING_ID = '${obj?.id}'")

    }

    override fun doDelete(condition: String?): Boolean {
        return genericDoDelete(Rating.RATING, condition, source)
    }
}